homeassistant/components/recorder/models.py

```diff
@@ -748,3 +746,48 @@ def __eq__(self, other: Any) -> bool:
             and self.state == other.state
             and self.attributes == other.attributes
         )
+
+
+def decode_attributes_from_row(
+    row: Row, attr_cache: dict[str, dict[str, Any]]
+) -> dict[str, Any]:
+    """Decode attributes from a database row."""
+    source: str = row.shared_attrs or row.attributes
+    if (attributes := attr_cache.get(source)) is not None:
+        return attributes
+    if not source or source == EMPTY_JSON_OBJECT:
+        return {}
+    try:
+        attr_cache[source] = attributes = json.loads(source)
+    except ValueError:
+        _LOGGER.exception("Error converting row to state attributes: %s", source)
+        attr_cache[source] = attributes = {}
+    return attributes
+
+
+def row_to_compressed_state(
+    row: Row,
+    attr_cache: dict[str, dict[str, Any]],
+    start_time: datetime | None = None,
+) -> dict[str, Any]:
+    """Convert a database row to a compressed state."""
+    if start_time:
+        last_changed = last_updated = start_time.timestamp()
+    else:
+        row_changed_changed: datetime = row.last_changed
+        if (
+            not (row_last_updated := row.last_updated)
+            or row_last_updated == row_changed_changed
+        ):
+            last_changed = last_updated = process_datetime_to_timestamp(
+                row_changed_changed
+            )
+        else:
+            last_changed = process_datetime_to_timestamp(row_changed_changed)
+            last_updated = process_datetime_to_timestamp(row_last_updated)
+    return {
+        COMPRESSED_STATE_STATE: row.state,
+        COMPRESSED_STATE_ATTRIBUTES: decode_attributes_from_row(row, attr_cache),
+        COMPRESSED_STATE_LAST_CHANGED: last_changed,
+        COMPRESSED_STATE_LAST_UPDATED: last_updated,
+    }
```

Comment:
This will always include all changes. We could consider adopting the format we used for `subscribe_entities` and only update the fields that have changed. Or is calculating that too much for that many objects?

Reply:
Since most high volume entities (ie power sensors) are requested with `minimal_responses` we collapse duplicate states anyways https://github.com/home-assistant/core/pull/71688/files#diff-b02f0838026f1d166cc53da13ba78e3860348a33e5ae05546d2e51e1a7a214f6R717

It might make a difference for climate entities but probably not worth optimizing for that since they don't tend to be that chatty compared to energy sensors.